base_ical/readme/CONFIGURE.rst

```diff
@@ -0,0 +1,10 @@
+To configure this module, you need to:
+
+#. Go to Settings/Technical/iCalendars
+#. Create a calendar, fill in the model you want to expose and possibly a domain to restrict records. You can use the ``user`` variable to restrict things relative to the user using the calendar
+#. A few iCalendar-fields have defaults that should work for any model, you'll have to fill in expressions manually though for the start and end date of the records.
+
+   For example, for model ``calendar.event``, you'd fill in ``record.allday and record.start_date or record.start`` as `DTSTART` and ``record.allday and record.stop_date or record.stop`` as `DTEND`.
+
+   For model ``hr.leave``, you'd write ``(record.request_unit_half or record.request_unit_hours) and record.date_from or record.date_from.date()`` for `DTSTART` and ``(record.request_unit_half or record.request_unit_hours) and record.date_to or (record.date_to.date() + timedelta(days=1))`` for `DTEND` - this is a bit more complex because of the way Odoo handles the begin and end times of leaves, and you'll want the extra day as most clients interpret the end date as non-inclusive.
```

Comment:
This shows a bit of the disadvantage of the approach here because you need quite complex `or ... and ... or` constructs to achieve some stuff because you can't use variables in such snippets. For calendar.event it would be nice to use the core function `_get_ics_file` too. I would suggest to use a single multi-line snippet with a defined interface and use safe_eval in "exec" mode. The downside is that it's more complicated but more flexible and allows additional fields like `DESCRIPTION`, `RRULE` ...

```python
# Also add vobject to the context
context["events"] = self._get_events()
safe_eval.safe_eval(self.code, context, mode="exec", nocopy=True)

# We could define that result can either be a list of dict or an iCalendar object
result = context.get("result", None)

if isinstance(result, list):
    calendar = vobject.iCalendar()
    for event in result:
        vobj = cal.add("vevent")
        for key, val in event.items():
            # TZ check for datetime etc.
            vobj.add(key).value = val
    return calendar.serialize()
return result
```

Snippets could be:
- calendar.event: `result = events._get_ics_file()`
- hr.leave:
```python
result = []
for event in events:
    if event.request_unit_half or event.request_unit_hours:
        dtstart = event.date_from
        dtend = event.date_to
    else:
        dtstart = event.date_from.date()
        dtend = event.date_from.date()
    result.append(
        {
            "dtstart": dtstart,
            "dtend": dtend,
            "summary": event.name,
            "uid": ...,  # a bit unsure but I would just pass a function -> record to uid to the snippet
        }
    )
```

Funnily enough that would make it possible to generate `vtodo` based on `mail.activity` aswell.

Reply:
can't we have both? As in (pseudo-python):

```python
result = {
  'dtend': eval(expression_dtend),
  ...
}

if code_full:
  safe_eval(code_full, locals)
  result.update(**locals['result'])
```

and the field for code_full would go to `base.group_noone` or a dedicated one? Then for simple cases admins can use the fields, if you want to do anything fancy you switch to the full code mode.

I'm all for injecting some helper functions in the eval context, and vtodos, sure. I think I'd also implement this with a simple mode (radio for this is a vevent or vtodo, showing/hiding the appropriate expression fields) and the full code mode where you can overwrite all of the above.

Reply:
Both might be complicated because the `_get_ics_file` are generating full calendars => I would run the snippet on multiple events instead of one event which would make merging both hard.

Maybe I can go with a switch for either simple or advanced mode but mixing is complex. In the end it boils down to the administrator to set it up. We could also just offer an advanced mode but some snippets for various basic cases.